elf_reader.go

```diff
@@ -520,8 +516,48 @@ func mapSpecFromBTF(spec *btf.Spec, name string) (*MapSpec, error) {
 				return nil, fmt.Errorf("can't get BTF map max entries: %w", err)
 			}
 
+		case "key":
+			if keySize != 0 {
+				return nil, errors.New("both key and key_size given")
+			}
+
+			pk, ok := member.Type.(*btf.Pointer)
+			if !ok {
+				return nil, fmt.Errorf("key type is not a pointer: %T", member.Type)
+			}
+
+			key = pk.Target
+
+			size, err := btf.Sizeof(pk.Target)
+			if err != nil {
+				return nil, fmt.Errorf("can't get size of BTF key: %w", err)
+			}
+
+			keySize = uint32(size)
+
+		case "value":
+			if valueSize != 0 {
+				return nil, errors.New("both value and value_size given")
+			}
+
+			vk, ok := member.Type.(*btf.Pointer)
+			if !ok {
+				return nil, fmt.Errorf("value type is not a pointer: %T", member.Type)
+			}
+
+			value = vk.Target
+
+			size, err := btf.Sizeof(vk.Target)
+			if err != nil {
+				return nil, fmt.Errorf("can't get size of BTF value: %w", err)
+			}
+
+			valueSize = uint32(size)
+
 		case "key_size":
-			if _, isVoid := keyType.(*btf.Void); !isVoid {
+			// Key needs to be nil and keySize needs to be 0 for key_size to be
+			// considered a valid member.
+			if key != nil || keySize != 0 {
```

Comment:
Doesn't this always bail out if keySize != 0?

Reply:
Oh I see, that is kind of confusing. This is to avoid the void type assertion?

Reply:
Ah, no, that's because `key, value` are no longer parsed or initialized in `(*Spec) Map()`, so we no longer need to make the distinction between `nil` and `*Void` here. We also cannot assume that e.g. `key` and `key_size` will appear in a specific order in the map definition.

I realize now that it should be impossible for `keySize` to be > 0 if `key` is nil, but I figured it wouldn't hurt to check both, just in case some behaviour is changed later on.

Reply:
I think I'd prefer just the `key` check, it took me a minute or two to understand why `keySize != 0` doesn't trigger in this case.